kafkabp/consumer.go

```diff
@@ -0,0 +1,228 @@
+package kafkabp
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ConsumeMessageFunc is a function type for consuming consumer messages.
+type ConsumeMessageFunc func(ctx context.Context, msg *sarama.ConsumerMessage) error
+
+// ConsumeErrorFunc is a function type for consuming consumer errors.
+type ConsumeErrorFunc func(err error)
+
+// consumer is an instance of a Kafka consumer.
+type consumer struct {
+	cfg ConsumerConfig
+	sc  *sarama.Config
+
+	consumer           atomic.Value // sarama.Consumer
+	partitions         atomic.Value // []int32
+	partitionConsumers atomic.Value // []sarama.PartitionConsumer
+
+	closed          int64
+	consumeReturned int64
+	offset          int64
+
+	wg sync.WaitGroup
+}
+
+// Consumer defines the interface of a consumer struct.
+type Consumer interface {
+	io.Closer
+
+	Consume(ConsumeMessageFunc, ConsumeErrorFunc) error
+
+	// IsHealthy returns false after Consume returns.
+	IsHealthy() bool
+}
+
+// NewConsumer creates a new Kafka consumer.
+func NewConsumer(cfg ConsumerConfig) (Consumer, error) {
+	sc, err := cfg.NewSaramaConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate input parameters.
+	if len(cfg.Brokers) == 0 {
+		return nil, ErrBrokersEmpty
+	}
+
+	if cfg.Topic == "" {
+		return nil, ErrTopicEmpty
+	}
+
+	kc := &consumer{
+		cfg:    cfg,
+		sc:     sc,
+		offset: sc.Consumer.Offsets.Initial,
+	}
+
+	// Initialize Sarama consumer and set atomic values.
+	if err := kc.reset(); err != nil {
+		return nil, err
+	}
+
+	return kc, nil
+}
+
+func (kc *consumer) getConsumer() sarama.Consumer {
+	c, _ := kc.consumer.Load().(sarama.Consumer)
+	return c
+}
+
+func (kc *consumer) getPartitions() []int32 {
+	p, _ := kc.partitions.Load().([]int32)
+	return p
+}
+
+func (kc *consumer) getPartitionConsumers() []sarama.PartitionConsumer {
+	pc, _ := kc.partitionConsumers.Load().([]sarama.PartitionConsumer)
+	return pc
+}
+
+// reset recreates the consumer and assigns partitions.
+func (kc *consumer) reset() error {
+	if c := kc.getConsumer(); c != nil {
+		if err := c.Close(); err != nil {
+			log.Warnw("Error closing the consumer", "err", err)
```

Comment:
For a truly library code calling `log.Warnw` directly is not recommended. Looks like this is the only instance direct log call is used in the whole PR, so I'd recommend make it a `log.Wrapper` field of the config struct.

Reply:
```suggestion
			return err
```

Any qualms to just returning the error? It seems like this is more in line with the philosophy outlined in `../log/wrapper.go`.